secure-api-gateway-ig-extensions/src/main/java/com/forgerock/sapi/gateway/dcr/request/RequestAndSsaSignatureValidationFilter.java

```diff
@@ -0,0 +1,319 @@
+        this.registrationRequestObjectFromJwtSupplier = registrationRequestObjectFromJwtSupplier;
+        this.supportedSigningAlgorithms = supportedSigningAlgorithms;
+        this.ssaValidator = ssaValidator;
+        this.registrationRequestJwtValidator = registrationRequestJwtValidator;
+        this.dcrUtils = dcrUtils;
+        log.debug("RequestAndSsaSignatureValidationFilter constructed");
+    }
+
+    @Override
+    public Promise<Response, NeverThrowsException> filter(Context context, Request request, Handler next) {
+        final String fapiInteractionId = FAPIUtils.getFapiInteractionIdForDisplay(context);
+        log.debug("({}) filtering - filtering out invalid signatures of registration request and ssa jwts",
+                fapiInteractionId);
+        if (!VALIDATABLE_HTTP_REQUEST_METHODS.contains(request.getMethod())) {
+            return next.handle(context, request);
+        }
+
+        try {
+            SignedJwt registrationRequestJwt = getRegistrationRequestObjectOrThrow(fapiInteractionId, context, request);
+            checkJwtSigningAlgorithmIsValid(fapiInteractionId, registrationRequestJwt, supportedSigningAlgorithms);
+            final JwtClaimsSet registrationRequestJwtClaimsSet = registrationRequestJwt.getClaimsSet();
+
+            final String ssaJwtString = getSsaB64EncodedJwtString(fapiInteractionId, registrationRequestJwtClaimsSet);
+            log.debug("({}) ssa from registration request jwt is {}", fapiInteractionId, ssaJwtString);
+            SignedJwt ssaSignedJwt = getSoftwareStatementAssertionSignedJwt(fapiInteractionId, ssaJwtString, supportedSigningAlgorithms);
+            final JwtClaimsSet ssaClaimsSet = ssaSignedJwt.getClaimsSet();
+
+            Promise<Response, DCRSignatureValidationException> ssaValidationPromise =
+                    ssaValidator.validateSoftwareStatementAssertionSignature(fapiInteractionId, ssaSignedJwt);
+            return ssaValidationPromise.thenAsync( response ->
+                registrationRequestJwtValidator.validateRegistrationRequestJwtSignature(
+                            fapiInteractionId, ssaClaimsSet, registrationRequestJwt)
+                            .thenAsync(regRequestValidationResponse -> {
+                        if (regRequestValidationResponse.getStatus() != Status.OK) {
+                            return Promises.newResultPromise(response);
+                        }
+                        return next.handle(context, request);
+                    }, ex -> {
+                        return Promises.newResultPromise(new Response(Status.BAD_REQUEST).setEntity(getJsonResponseBody(ex)));
+                    }), ex -> {
+                        log.debug("({}) Failed to get jwks from jwks_uri {}", fapiInteractionId, "caught exception: " + ex.getMessage(), ex);
+                        return Promises.newResultPromise(new Response(Status.BAD_REQUEST));
```

Comment:
Review exception handling and logging here.

If you switch to using the com.forgerock.sapi.gateway.dcr.ValidationException then you can also make use of com.forgerock.sapi.gateway.dcr.ErrorResponseFactory to build the response objects which are OAuth 2.0 compliant, currently the response has an empty body.